### DISCUSSION TITLE
Glide mode

### DISCUSSION THREAD

**Discussion by brats2017** [JSBSim-Team/jsbsim#385] [Support requests]
Hello all
I have a question and it will be appreciated if you help me with this. I wanted to ask you is implemented the boost-glide mode for an aircraft in JSBSim? If yes, in which aircraft code I can find it? I know the gliding mode is implemented for X-15. However, what about boost-glide mode?
Thanks in advance.

**Comment by seanmcleod** [comment]
What do you mean by 'boost-glide mode'?

Do you mean like the X-15 which get's dropped from the wing of a B-52 carrier aircraft, then lights it's rocket engine for a couple of minutes, which is the boost? Then after the engine cuts out it glides back down?

**Comment by brats2017** [comment]
Thank you for your response @seanmcleod. I understood what is the boost mode in X-15 with your explanation. Just one question: is there any possibility/idea to increase the duration of boost mode in X-15?

**Comment by seanmcleod** [comment]
But you were asking about 'boost-glide mode', so let's first make sure we're talking about the same thing. Do we agree on what 'boost-glide mode' means?

If you want to increase the duration of the boost mode of the X-15 simply increase the amount of rocket fuel available.

**Comment by brats2017** [comment]
The boost-glide mode that is in my mind is the same as discussed in this [paper](https://www.researchgate.net/publication/315833583_Trajectory_optimisation_for_a_rocket-assisted_hypersonic_boost-glide_vehicle).

> If you want to increase the duration of the boost mode of the X-15 simply increase the amount of rocket fuel available.

Thank you for your helpful idea @seanmcleod.

**Comment by brats2017** [comment]
We have four phases in the [paper](https://www.researchgate.net/publication/315833583_Trajectory_optimisation_for_a_rocket-assisted_hypersonic_boost-glide_vehicle), and these phases are described below and depicted in Fig. 1.
1. The first 5 seconds of first stage boost phase, during which pitch manoeuvre does not take place. In this phase, the vehicle gains sufficient speed in the first 5 seconds, after which the stabilisers become effective and provide sufficient damping moment for a controlled pitch manoeuvre(17).
2. The time duration after t = 5s till the burn-out of the first stage rocket motor. It is during this phase that the launch vehicle pitches down using angle-of-attack control.
3. The second stage boost phase. In this phase, the flight path angle changes to meet the burn-out conditions.
4. The free flight and the re-entry stage. In this phase, the vehicle is steered to optimal down-range or cross-range distance with the help of angle-of-attack, bank angle and throttle control.
![boost-glige](https://user-images.githubusercontent.com/65332612/106150457-603c3f00-6149-11eb-8e18-357d4026cf14.JPG)

-I think the boost-glide hypersonic that is in the above paper, is launched from a rocket while the boost-glide of X-15 in JSBSim is launched from another aircraft. Do you think am I right?

-My other question is do you think it is reasonable that X-15 is launched from a rocket?
IF yes, is there any code for the boost-glide modes in JSBSim that I can use for the lunch of X-15 from a rocket? Do you think the J2461.xml can help me with this issue?

Your help will be greatly appreciated.
Thanks in advance

**Comment by seanmcleod** [comment]
It's very difficult to provide useful responses without a detailed description of what you're trying to implement. So just asking about "boost-glide" in terms of whether it's been implemented (depending on exactly what it actually means, and without defining it clearly) in an existing JSBSim aircraft doesn't provide nearly enough information.

First of all reading through the paper it turns out that they test 2 different aircraft, a wave-rider aircraft and a wing-body aircraft, and 2 different configurations of the wing-body aircraft, one with no engine and one with a rocket engine. So 3 different configurations in all. 

These are all launched using a rocket. The wave-rider aircraft and the wing-body aircraft with no engine are pure gliders once they separate from the rocket that launched them. The wing-body aircraft uses a combination of gliding and can use it's rocket engine for short periods.

So first of all, what exactly are you trying to model. Do you want to model the complete system from launch on the ground and model all 3 configurations? 

Or do you only want to model say 1 of the configurations from the time it detaches from the rocket until it lands?

How closely do you want to model their results? If you want to be very close then you have to use their aerodynamic model etc., you can't just make use of the X-15 in it's place and expect to get results that are close to theirs.

**Comment by brats2017** [comment]
Thank you so much for reading the paper and your explanation. 

> Or do you only want to model say 1 of the configurations from the time it detaches from the rocket until it lands?

I had sent the paper to clarify my think related to boost-glide mode. However, it is not important that I modeled the configuration exactly that is defined in the paper.
Consequently, right now, please do not consider the paper that I sent, and please let me know what you think about my following question:

I just wanted to change the configuration of the x153.xml file. The only thing that I want to change is related to the launch phase of the aircraft. How can I ask X-15 to be launched from a rocket instead of being launched from an aircraft (B-52)? 

Please consider that I am a beginner with aerospace and JSBSim.
Your help will be great;y appreciated.
Thanks in advance.

**Comment by seanmcleod** [comment]
I think you're over complicating things in terms of thinking about a rocket versus a carrier/mothership like a B-52. All you really care about are the **initial conditions** at launch, you can set the initial conditions as you require them, doesn't matter what the mothership was.

So take a look at `scripts\x153.xml`, you'll notice the following:

```xml
  <use aircraft="X15" initialize="reset01"/>
``` 

The **initial conditions** are specified in the file **reset01** which lives in `aircraft\x15\reset01.xml`, here is a copy.

```xml
<?xml version="1.0"?>
<initialize name="reset01">
  <!--
    This file sets up the X15 to start off at altitude.
  -->
  <ubody unit="FT/SEC">  750.0  </ubody>
  <vbody unit="FT/SEC">    0.0  </vbody>
  <wbody unit="FT/SEC">  100.0  </wbody>
  <latitude unit="DEG">   33.0  </latitude>
  <longitude unit="DEG">-110.0  </longitude>
  <phi unit="DEG">         0.0  </phi>
  <theta unit="DEG">       0.0  </theta>
  <psi unit="DEG">       260.0  </psi>
  <altitude unit="FT"> 40000.0  </altitude>
</initialize>
```

So it starts the simulation off with the aircraft at 40,000ft, at a particular lat, lon, particular attitude and particular velocity.

So if you want to start the simulation off as if the X-15 had just been released from a rocket then create a new **initial conditions** file and you can for example set the starting altitude to 200,000ft, with a much faster starting velocity, and a different attitude, e.g. maybe it gets released from the rocket at a pitch attitude of 60 degrees etc.

**Comment by brats2017** [comment]
Thank you so much for your response. I did as you mentioned and I got the results that I want.

Also, about the paper that I sent and according to the guidance that you gave me, I used the aerodynamic model of waverider configuration for re-entry flight (phase 4 in the figure above). It behaves as a glider once it separates from the rocket that launched it. However, right now, I want to model the complete system in JSBSim (phase 1, phase 2,  phase 3, and phase 4).

I think I should define 2 different strategies for that:
To model phase1, phase2, and phase3: I should use the aerodynamic forces that exist in the main version of the X15.xml file. 
To model phase 4: I should use the aerodynamic forces that I defined for waverider and exist in the paper.
 
Can you please let me know do you agree with this? If yes, How can I do that?

I know that the events defined in the main file of x153.xml consist of these phases with another strategy. However, I do not know how can I model phases 1,2, and 3 (boost modes) when I want to model a waverider configuration for the gliding/reentry mode.

Your help will be greatly appreciated.
Thanks in advance

**Comment by seanmcleod** [comment]
If you want to model phases 1, 2 & 3 then you need to define a rocket model, it's aerodynamic model will be quite different to an X-15. Plus it gets much more complicated in that during after separation you then have to switch models completely in terms of the mass, moments of inertia, aerodynamic model etc.

Take a look at the Space Shuttle model available for FlightGear since they have to model these sorts of changes.

If the Wave-Rider is housed inside a fairing then it's aerodynamic model mustn't be active during the launch phases, but if it sits on the rocket without a fairing covering it then it's aerodynamic model in addition to the aerodynamics of the rocket itself have to be taken together, the Space Shuttle falls into this category.

**Reply by brats2017** [reply]
Thank you so much for your response. Your explanation is very helpful.
Right now, I am close to the deadline to get results, so I want to do the easier way to get an initial answer. I need some clarification to continue my project.

My question is related to your comments:

> Take a look at the Space Shuttle model available for FlightGear since they have to model these sorts of changes.

I ran the j2461.xml file and I extend the runtime and got the following result. I do not know why it is dropped, but I think I can use this for my gliding mode. My question is do you think, can I ask the space shuttle to carry the X-15 and drop it (using the model of x-15 that we already have, and regardless of the wave-rider configuration)?
![J246](https://user-images.githubusercontent.com/65332612/106299812-a82f9480-6223-11eb-8dac-8dd5bf436c2e.JPG)

> If the Wave-Rider is housed inside a fairing then it's aerodynamic model mustn't be active during the launch phases.

It will be good if I could implement that because in that case, I do not have to switch the aerodynamics forces. Do you have any idea that how can I implement that?


Overall, what is your suggestion regarding the modeling of all 4 phases of boost-glide according to my limited time?
Your help will be greatly appreciated.
Thanks in advance

**Reply by bcoconni** [reply]
The space shuttle model is very complicated and I guess you'd better use the J246 rocket model that is in JSBSim. The model is much simpler and I guess it could be adapted to your needs.

In the `J2461.xml` script that you mentioned above, stage separation is modelled by simply setting the masses of the jettisoned stages to zero. See for example the code below: it simulates the SRBs (i.e. Solid Rocket Boosters) jettison by setting the properties `inertia/pointmass-weight-lbs[0]` and `inertia/pointmass-weight-lbs[1]` to 0.
https://github.com/JSBSim-Team/jsbsim/blob/261f708b9d9bdc5828064da923d6d5378b48cf65/scripts/J2461.xml#L132-L140
Merging the `J246` model with the `X-15` will not be straightforward but I think it is definitely feasible.

**Reply by brats2017** [reply]
Thank you so much for your reply @bcoconni. 
I think it is a good solution for the separation stage. However, I think the problem will not be solved because the aerodynamic forces for x15 are different in glide mode (after separation) and should be switched to new aerodynamic forces. Do you agree with it?
Thanks in advance

**Reply by bcoconni** [reply]
> I think the problem will not be solved because the aerodynamic forces for x15 are different in glide mode (after separation) and should be switched to new aerodynamic forces.

Given your tight schedule, I guess you have only 2 options:
1. Assume that the X-15 is protected by some aerodynamic fairings during the launch and until the X15 is separated from the rocket. In that case, you would switch from the J246 aerodynamics to the X15 aerodynamics just after the X15 has been separated from the rocket.
2. Build an aerodynamic model that assumes that the X15 aerodynamics are only slightly modified by the rocket (say the drag along X is equal to the X15 drag and the drag in Y & Z directions is the sum of the rocket and the X15 drag). Or something like that.

**Reply by brats2017** [reply]
Thank you so much for your great solutions. However, I need more time to implement these options. I'll definitely work on them in the future.

**Comment by bcoconni** [comment]
This issue (#383) has been moved to `Discussions`.

Support requests and questions are now resolved in [`Discussions`](https://github.com/JSBSim-Team/jsbsim/discussions) while [`Issues`](https://github.com/JSBSim-Team/jsbsim/issues) are restricted to bugs reports and feature requests.

This change is only motivated by project organization and will not affect the support provided. We'll gladly continue to answer your questions 😄.

**Comment by seanmcleod** [comment]
> deadline to get results

Who do you have to provide the results to?

> my limited time

Define limited. Are you talking a couple of days, or a couple of weeks?

I know of the Space Shuttle model, but I've never looked at the source of the model to see how complicated it is and how easy it would be to adapt.

**Reply by brats2017** [reply]
I have up to 3 days to get an initial result.

> I know of the Space Shuttle model, but I've never looked at the source of the model to see how complicated it is and how easy it would be to adapt.

Can you please let me know your idea about that after you had a chance to review the code?

Your help will be greatly appreciated.
Thanks in advance

**Reply by brats2017** [reply]
The only idea that is on top of my head is using the final status of J246 as an initial condition for x15 and running them separately and merging their .csv file to get results of boost-glide mode. Can you please let me know what is your idea about that? @seanmcleod @bcoconni 
Thanks in advance

**Reply by bcoconni** [reply]
That's definitely an option but for the sake of completeness, you need to modify the mass of the J246 payload to be equal to the X15 mass. Otherwise, you'll end up with a launch velocity that is not compatible with your assumptions.

**Reply by brats2017** [reply]
Thank you so much for your reply. I will surely consider it.
Thank you very much for your help @seanmcleod @bcoconni. This discussion was very helpful for me.

**Reply by seanmcleod** [reply]
Out of interest is this for a university course/project? Interested to know the context in which JSBSim is being used.

**Reply by brats2017** [reply]
It is in university.

**Comment by brats2017** [comment]
Hello all,
I want to ask you is it possible in JSBSim to run two configurations in one script?
To be more specific, I have two configurations that the first one starts from 0 seconds and ends at 200 seconds. The second simulation starts at 200 seconds and ends at 400 seconds which its initial conditions are the final conditions of the first simulation. I want to run these files consecutively in such a way the initial conditions of the second simulation be read automatically. Do you think is it possible with JSBSim or I have to do that out of JSBSim?
Thanks in advance

**Comment by jonsberndt** [comment]
Seems to me it is (or should be) possible to write an initial conditions file upon command, which is in turn read by the second script. Unfortunately, I'm about to fall asleep 😴  and will have to take a look at this tomorrow.
Jon

**Reply by brats2017** [reply]
Thanks for your reply @jonsberndt. Did you have a chance to look at that? I think I should write a C++ code upon JSBSim to read the parameters and files. Is there a sample code for this?

**Comment by seanmcleod** [comment]
What about something along these lines?

```python
import jsbsim

fdm = jsbsim.FGFDMExec('.', None)
fdm.load_script('scripts/rocketlaunch.xml')
fdm.run_ic()

while fdm.run():
  pass

# Grab final values from fdm run, position, attitude and velocities and 
# create an initial conditions file for the 2nd script
fdm['position/h-sl-ft']
...
...

fdm = jsbsim.FGFDMExec('.', None)
fdm.load_script('scripts/glide.xml')
fdm.run_ic()

while fdm.run():
  pass

```

**Reply by bcoconni** [reply]
@brats2017 , alternatively JSBSim can generate an initialization file from your current position and attitude with the property `simulation/write-state-file`. If you set this property to `1` or `2`, JSBSim will write a file `initfileXXXXX.xml` in your aircraft folder (where `XXXXX` is replaced with the simulation time at which the file is generated).

You most likely need an initialization file with the format v2.0 which contains more information about the aircraft position and attitude, in which case the property `simulation/write-state-file` must be set to `2`.

To answer your original answer, you can modify your first script to terminate by setting the property `simulation/write-state-file` to 2 and write a second script that read that `initfileXXXXX.xml` as an input. Then if you execute both of them in a row, you should get the execution sequence you are asking for.

With a shell command like `bash` this can be done with the following shell command:
```bash
> JSBSim my_first_script.xml && JSBSim scripts/my_second_script.xml
```
The `&&` operation guarantees that the second script will only be executed once the first script has succeeded.

**Reply by brats2017** [reply]
Thank you so much, @bcoconni for your detailed answer. 

> If you set this property to 1 or 2, JSBSim will write a file initfileXXXXX.xml

I did it and the initialization file has been created in the aircraft folder.

>   With a shell command like bash this can be done with the following shell command:
 JSBSim my_first_script.xml && JSBSim scripts/my_second_script.xml

Is there any idea on what is the syntax of this command for the Command Argument in the following window? I have an error in using this syntax: `−−script=scripts/my_first_script.xml && scripts/my_second_script.xml`

![Capture6](https://user-images.githubusercontent.com/65332612/115101407-ea5a9f00-9f11-11eb-9748-97b8c0460c78.JPG)

Thanks in advance.

**Reply by bcoconni** [reply]
I am no specialist of MSVC but I'm pretty sure you cannot do that. The syntax I have suggested must be run from a terminal. It consists in running the `JSBSim` executable twice in a row, something that is most likely not possible from the debugging options of MSVC.

The bash command ` JSBSim --script=scripts/B52.xml && JSBSim --script=scripts/x15.xml` is nothing more than executing the 2 commands in sequence:
```bash
> JSBSim --script=scripts/B52.xml
> JSBSim --script=scripts/x15.xml
```
The `&&` command is just some bash/terminal syntax sugar to avoid waiting for the first command to complete before running the next one.

**Reply by seanmcleod** [reply]
Correct, in the MSVC example effectively you're telling MSVC to execute the following each time you run the program:

`jsbsim.exe --script=scripts/B52.xml && scripts/x15.xml`

JSBSim doesn't have support for running multiple scripts, as @bcoconni mentioned you need to run two separate instances of JSBSim with the relevant script command one after the other. Either manually, or using the `&&` support from a shell like Bash or the Command Prompt shell in Windows, e.g.

![image](https://user-images.githubusercontent.com/824856/115112067-f0896380-9f83-11eb-85d5-670929d282d7.png)

**Reply by brats2017** [reply]
Thank you so much for your responses. One question is that is there any way to move the initialization file (related to the first aircraft which has been created with setting `simulation/write-state-file` to 2) automatically to the folder of the second aircraft? I copied it manually and I got the result that I wanted.

Thanks in advance

**Reply by bcoconni** [reply]
>  is there any way to move the initialization file (related to the first aircraft which has been created with setting simulation/write-state-file to 2) automatically to the folder of the second aircraft ?

I am afraid not. But if you are relying on shell/terminal commands, you could very well instruct the shell to copy the file from one folder to the other:
```bash
> JSBSim --script=scripts/B52.xml
> cp aircraft/B52/initfileXXX.xml aircraft/x15/.
> JSBSim --script=scripts/x15.xml
```

**Reply by brats2017** [reply]
Thank you @bcoconni.

**Comment by seanmcleod** [comment]
Didn't know about the `simulation/write-state-file` option. I see both option 1 and 2 output the pose (3D position and attitude) and velocity, but option 2 also outputs the angular velocities.

**Reply by bcoconni** [reply]
Well, hum, documentation is not our strong suit, is it ? :wink: 
> I see both option 1 and 2 output the pose (3D position and attitude) and velocity, but option 2 also outputs the angular velocities.

Yes option 2 outputs an initialization file in format v2.0
https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/src/models/FGPropagate.cpp#L775
Format v2.0 indeed allows to specify more parameters.

**Comment by seanmcleod** [comment]
> Well, hum, documentation is not our strong suit, is it ?

Part of the reason I monitor the issues/discussion lists and try answer questions, so I can learn by searching through the code 😉
